adventure-crew-game/Assets/Scripts/Combat/CombatEntity.cs

```diff
@@ -0,0 +1,146 @@
+    public float cooldown = 1.0f;
+    private float timer = 0.0f;
+    public enum CurrentAction
+    {
+        idle,
+        moving,
+        attacking
+    }
+    public CurrentAction currentAction = CurrentAction.idle;
+
+    void Start()
+    {
+        //tag check
+        if (CompareTag("Adventurer") == false && CompareTag("Enemy") == false) Debug.LogWarning("This entity is set to the wrong tag");
+    }
+
+    public void DecideCombatAction()
+    {
+        currentAction = CurrentAction.idle;
+        if (target == null)
+        {
+            if(FindTarget()) DecideCombatAction();
+            //if can't find any opponent, then the combat is over
+            else CombatManager.Instance.CombatOver();
+        }
+        else // move or attack
+        {
+            //Debug.Log("Distance: " + (target.transform.position - transform.position).magnitude);
+
+            if ((target.transform.position - transform.position).magnitude > stats.Range) 
+            {
+                Debug.Log(gameObject.name + "'s action should be moving");
+                currentAction = CurrentAction.moving; 
+            }
+
+            else currentAction = CurrentAction.attacking;
+        }
+        
+    }
+
+    virtual public bool FindTarget()
+    {
+        target = CombatManager.Instance.GetOpponent(transform.position, tag);
+        return target != null;
+        //if we are going to add a healer, we can override this method
+        //and pass in the Enemy tag, so the healer find the target among adventurers
+    }
+    
+
+    private bool MoveCheck()
+    {
+        if (target == null)
+        {
+            DecideCombatAction();
+            return false;
+        }
+        if ((target.transform.position - transform.position).magnitude < stats.Range)
+        {
+            DecideCombatAction();
+            return false;
+        }
+        return true;
+    }
+    public void Move(int agility)
+    {
+        //Debug.Log(gameObject.name + " Moving");
+        Vector3 moveDir = (target.transform.position - transform.position).normalized;
+        transform.position += moveDir * agility * Time.deltaTime;
+        transform.LookAt(target.transform.position);
+    }
+
+    private bool AttackCheck()
+    {
+        if (target == null)
+        {
+            DecideCombatAction();
+            return false;
+        }
+        if ((target.transform.position - transform.position).magnitude > stats.Range)
+        {
+            DecideCombatAction();
+            return false;
+        }
+        return true;
+    }
+    public void Attack(int damage)
+    {
+        if (timer <= 0.0) 
+        {
+            timer = cooldown;
+            target.TakeDamage(damage);
+
+            //visual
+            TurnToTarget(target.transform.position);
+        }
+        else
+        {
+            timer -= Time.deltaTime;
+        }
+    }
+
+    public void TakeDamage(int damage)
+    {
+        Debug.Log("Got damage " + damage);
```

Comment:
Please clean up after your Debug.Logs once you're done using them

Reply:
I'm going to keep that for a little longer, once I implement health bar, I will clean that up. Thanks for reminding me!